2_process/src/add_dynamic_attr.R

```diff
@@ -0,0 +1,796 @@
+  #' @param lag_table table with 3 columns: attributes, lags, and lag_unit.
+  #' This is used to compute the lagged attributes. Can set to NULL to use a 
+  #' lag of 0.
+  #' @param attr_years ordered character vector of the years available for this 
+  #' dataset. Should be in ascending order.
+  #' In attrs, for each attr_prefix, there should be one column for each year.
+  #' @param date_ranges Date vector that specifies the transition dates from one
+  #' year to another year. Dates in dyn_df less than the first date 
+  #' will be assigned to the first year in attr_years. Dates in dyn_df greater 
+  #' than the last date will be assigned to the last year in attr_years.
+  #' @param attr_name name to use for this attribute. Format of the new name will
+  #' be paste0(attr_prefix, '_', attr_name, '_', lags, lag_unit)
+  #' @param attr_prefix character vector of prefixes for the attributes 
+  #' (e.g., CAT, TOT)
+  #' 
+  #' @return dyn_df with the new dynamic attribute columns.
+  
+  if (is.null(lag_table)){
+    lags <- 0
+    lag_unit <- ''
+  }else{
+    lags <- lag_table %>% 
+      filter(attribute == attr_name) %>%
+      pull(lags) %>%
+      unlist()
+    lag_unit <- lag_table %>% 
+      filter(attribute == attr_name) %>%
+      pull(lag_unit) %>%
+      unlist()
+  }
+  
+  #Make the length of the lag_unit match the length of lags
+  if ((length(lag_unit) == 1) & (length(lags) > 1)){
+    lag_unit <- rep(lag_unit, length(lags))
+  }
+  
+  #format for new variable is NAME_lag
+  for (j in 1:length(attr_prefix)){
+    tmp_colnames <- paste0(attr_prefix[j], '_', attr_name, '_', lags, lag_unit)
+    dyn_df[tmp_colnames] <- NA_real_
+  }
+  rm(j, tmp_colnames)
+  
+  #Loop over all of the lags
+  for (lag in 1:length(lags)){
+    #Create a temporary date column containing the lagged date to use
+    if(lags[lag] != 0){
+      dyn_df$lagged_Date <- dyn_df$Date %m-% period(paste(lags[lag], lag_unit[lag]))
+    }else{
+      dyn_df$lagged_Date <- dyn_df$Date
+    }
+    
+    #Join data to all segs by date range
+    # Note: using this method instead of a case_when because
+    # case_when was very slow
+    for (i in 1:(length(date_ranges)+1)){
+      for (j in 1:length(attr_prefix)){
+        if (i == 1){
+          dyn_df[dyn_df$lagged_Date <= date_ranges[i], 
+                 c(paste0(attr_prefix[j], '_', attr_name,
+                          '_', lags[lag], lag_unit[lag]))] <- 
+            left_join(dyn_df %>% 
+                        filter(lagged_Date <= date_ranges[i]) %>% 
+                        select(seg), 
+                      attrs %>% 
+                        select(PRMS_segid, 
+                               paste0(attr_prefix[j], '_', attr_years[i])),
+                      by = c('seg' = 'PRMS_segid')) %>%
+            select(paste0(attr_prefix[j], '_', attr_years[i]))
+        }else if (i == (length(date_ranges)+1)){
+          dyn_df[dyn_df$lagged_Date > date_ranges[i-1], 
```

Comment:
I'll add a comment for this. It's there because the `date_ranges` specify transition dates between statis attribute datasets. So the `date_ranges` vector is 1 element shorter than the total number of static attribute datasets that are made into a dynamic timeseries